add best practices on how to get near real-time data with dbt (#8478)
fishtown-analytics/docs.getdbt.com: website/docs/best-practices/how-we-handle-real-time-data/5-views-only-pattern.md

---
title: "Views-only pattern for maximum freshness"
id: "5-views-only-pattern"
description: Preserve source table latency with lightweight view transformations
hoverSnippet: Learn when views-only patterns provide maximum data freshness
---

:::info Snowflake examples ahead

> This page uses Snowflake for code examples, but the views-only pattern can be adapted to other warehouses.

Reviewer: ```suggestion
This page uses Snowflake for code examples, but you can adapt the views-only pattern to other warehouses.
```

:::

For some workloads, the simplest and most "real-time" pattern is to materialize everything as views on top of a continuously updated source table. When transformations are very lightweight and the source is already being updated in near real-time, this can preserve the source's latency almost perfectly.

> #### When to use the views-only pattern

Reviewer: ```suggestion
## When to use the views-only pattern
```

Use this when:

- Source freshness is already "good enough" (for example, ingestion service or operational system writes into a warehouse table every few seconds or minutes).

> - You have very lightweight transformations:

Reviewer: ```suggestion
- You have very lightweight transformations, such as:
```

>   - Simple projections / renames.

Reviewer: ```suggestion
  - Simple projections / renames
```

>   - 1–2 joins to small reference tables.

Reviewer: ```suggestion
  - One to two joins to small reference table
```

>   - Minimal or no heavy aggregations / window functions.

Reviewer: ```suggestion
  - Minimal or no heavy aggregations or window functions
```

- You care most about preserving the source table's latency and are willing to trade off some query performance at read time.
- This works best for small-to-medium tables with simple queries.

Typical examples:

- Dashboards showing current system status (like active user sessions, current queue depth, or recent device heartbeats) where you need to see the latest data immediately.
- Event data that you're forwarding to other tools with minimal transformation (raw data with just a bit of normalization, like cleaning up field names or adding a few reference fields).

If your transform logic becomes heavier, multiple teams depend on the data, or you need better cost and performance control, transition to [incremental models](/best-practices/how-we-handle-real-time-data/2-incremental-patterns) or [dynamic tables/materialized views](/best-practices/how-we-handle-real-time-data/3-warehouse-native-features). Reserve this pattern for the smallest, most latency‑sensitive use cases.

#### Assumptions

> 

Reviewer: Can we also add an intro here?
```suggestion
The examples used in this page assume the following setup:
```

> - A non‑dbt system (ETL, streaming pipeline, app) is already writing into a warehouse‑resident table such as `raw.realtime_events` or `ops.active_sessions`, and that table meets your SLA for latency.

Reviewer: ```suggestion
- A non‑dbt system (ETL, streaming pipeline, app) is already writing into a warehouse‑resident table such as `raw.realtime_events` or `ops.active_sessions`, and that table meets your service level agreement for latency.
```

- Querying that table directly is acceptable from a performance and cost standpoint for your expected concurrency.
- You don't need dbt to persist intermediate tables; you mainly care about:

>   - Consistent SQL logic (column naming, type casting).

Reviewer: ```suggestion
  - Consistent SQL logic (column naming, type casting)
```

>   - Tests, contracts, and lineage.

Reviewer: ```suggestion
  - Tests, contracts, and lineage
```

>   - Exposures to BI / downstream tools.

Reviewer: ```suggestion
  - Exposures to BI / downstream tools
```

        loaded_at_field: event_ts
```

### Staging view

```sql
-- models/staging/stg_realtime_events.sql
{{ config(
    materialized = 'view'
) }}

select
    event_id,
    event_ts::timestamp_ntz   as event_ts,  -- Snowflake syntax for type casting
    to_date(event_ts)         as event_date,
    user_id,
    event_type,
    payload
from {{ source('raw', 'realtime_events') }};
```

### Domain view definition

```sql
-- models/marts/vw_realtime_events_enriched.sql
{{ config(
    materialized = 'view'
) }}

with base as (
    select *
    from {{ ref('stg_realtime_events') }}
),

user_dim as (
    select
        user_id,
        user_segment,
        signup_date
    from {{ ref('dim_user') }}   -- can be a table or incremental model
)

select
    b.event_id,
    b.event_ts,
    b.event_date,
    b.user_id,
    u.user_segment,
    b.event_type,
    b.payload
from base as b
left join user_dim as u
  on b.user_id = u.user_id;
```

Downstream tools query `analytics.vw_realtime_events_enriched`. As long as `raw.realtime_events` is continuously updated, this view stack is as fresh as the source.

## Benefits

> - Maximum freshness &mdash; The view reflects new data as soon as it lands in `raw.realtime_events`.

Reviewer: Suggesting to use colons here as well!

```suggestion
- Maximum freshness: The view reflects new data as soon as it lands in `raw.realtime_events`.
```

> - Simple operations &mdash; No incremental logic to tune and no extra dbt job needed just to keep the data fresh. You still schedule jobs for tests, docs, and so on.

Reviewer: ```suggestion
- Simple operations: No incremental logic to tune and no extra dbt job needed just to keep the data fresh. You still schedule jobs for tests, docs, and so on.
```

> - Best for small datasets &mdash; Works well when tables are small and queries are simple. Computing the view on the fly is cheap and fast.

Reviewer: ```suggestion
- Best for small datasets: Works well when tables are small and queries are simple. Computing the view on the fly is cheap and fast.
```

## Limitations and risks

This pattern is only safe under tight constraints and has several important limitations:
<!-- no toc -->
    - [Doesn't scale to heavy transformations](#doesnt-scale-to-heavy-transformations)
    - [No "frozen" intermediate tables](#no-frozen-intermediate-tables)
    - [Schema change sensitivity](#schema-change-sensitivity)
    - [Potential impact on operational systems](#potential-impact-on-operational-systems)

> #### Doesn't scale to heavy transformations

Reviewer: ```suggestion
### Doesn't scale to heavy transformations
```

If your logic evolves into large joins, deep view chains, or expensive aggregations, you'll quickly run into performance issues:

- Every query must re‑execute all the logic.
- The warehouse has to optimize and execute the full stack of views every time.

> In those cases, prefer:

Reviewer: ```suggestion
In those cases, use either of the following:
```

> - [Incremental models](/best-practices/how-we-handle-real-time-data/2-incremental-patterns), or

Reviewer: ```suggestion
- [Incremental models](/best-practices/how-we-handle-real-time-data/2-incremental-patterns)
```

> - [Dynamic tables or materialized views](/best-practices/how-we-handle-real-time-data/3-warehouse-native-features) where appropriate.

Reviewer: ```suggestion
- [Dynamic tables or materialized views](/best-practices/how-we-handle-real-time-data/3-warehouse-native-features), where appropriate
```

> #### No "frozen" intermediate tables

Reviewer: ```suggestion
### No "frozen" intermediate tables
```

Because everything is a view:

- There's no persisted intermediate layer to debug or profile.
- You can't easily "rerun yesterday's logic" if upstream data changes—everything always reflects the current state.

> #### Schema change sensitivity

Reviewer: ```suggestion
### Schema change sensitivity
```

> Schema changes in the source table propagate immediately through the view stack:

Reviewer: ```suggestion
Schema changes in the source table propagate immediately through the view stack, which:
```

- Can break downstream BI if columns are dropped or types change.

> - Tests and model contracts become more important, since there's no batch boundary to catch issues before users see them.

Reviewer: ```suggestion
- Make tests and model contracts more important, since there’s no batch boundary to catch issues before users see them.
```

> #### Potential impact on operational systems

Reviewer: ```suggestion
### Potential impact on operational systems
```

> If the continuously‑updated source is itself a live operational store (not a warehouse landing table), you must be careful not to overload it with analytics queries. In most cases, we recommend:

Reviewer: ```suggestion
If the continuously‑updated source is itself a live operational store (not a warehouse landing table), you must be careful not to overload it with analytics queries. In most cases, it is recommended to:
```

> - Replicating into a warehouse table first (Snowflake, BigQuery, Databricks, and so on), then

Reviewer: Suggesting to use numbers here since these should be done in the correct sequence
```suggestion
1. Replicate into a warehouse table first (Snowflake, BigQuery, Databricks, and so on).
```

> - Applying this views‑only pattern within the warehouse, not directly on the OLTP system.

Reviewer: ```suggestion
2. Apply this views‑only pattern within the warehouse, not directly on the Online Transaction Processing system.
```
